src/backend/distributed/connection/connection_management.c

```diff
@@ -445,15 +603,102 @@ ShutdownConnection(MultiConnection *connection)
 void
 FinishConnectionListEstablishment(List *multiConnectionList)
 {
+	int connectionIndex = 0;
+	int connectionCount = 0;
+	MultiConnectionReady *connections = NULL;
 	ListCell *multiConnectionCell = NULL;
 
+	WaitEventSet *waitEventSet = NULL;
+	bool waitEventSetRebuild = false;
+	int waitCount = 0;
+	WaitEvent *events = NULL;
+
+	/*
+	 * Here we prepare an array that keeps the MultiConnection together with the
+	 * information on if and how to wait before PQconnectionPoll should be called on it
+	 */
+	connectionCount = list_length(multiConnectionList);
+	connections = (MultiConnectionReady *) palloc0(connectionCount *
+												   sizeof(MultiConnectionReady));
 	foreach(multiConnectionCell, multiConnectionList)
 	{
-		MultiConnection *multiConnection = (MultiConnection *) lfirst(
-			multiConnectionCell);
+		MultiConnection *connection = (MultiConnection *) lfirst(multiConnectionCell);
+
+		connections[connectionIndex].connection = connection;
+
+		/*
+		 * before we can build the waitset to wait for asynchronous IO we need to know the
+		 * pollmode to use for the sockets. This is populated by executing one round of
+		 * PQconnectPoll. This updates the MultiConnectionReady struct with the ready state
+		 * and its next poll mode.
+		 */
+		MultiConnectionReadyExecutePoll(&connections[connectionIndex]);
+
+		if (!connections[connectionIndex].ready)
+		{
+			waitCount++;
+		}
+
+		connectionIndex++;
+	}
 
-		/* TODO: consider making connection establishment fully in parallel */
-		FinishConnectionEstablishment(multiConnection);
```

Comment:
Can we replace `FinishConnectionEstablishment` with the logic in this function?

Reply:
e.g. do `FinishConnectionListEstablishment(list_make1(connection));`

Reply:
I was thinking how to do this. Your suggestion makes a lot of sense, running tests with this now locally. Will push after they pass.